pkg/socketservice/socketservice.go

```diff
@@ -144,33 +147,53 @@ func deployLoop() {
 		}
 
 		for _, a := range apps {
-			if err := processDeploySocketForApp(clusterSocket, a); err != nil {
-				logger.Error(errors.Wrapf(err, "failed to run deploy loop for app %s in cluster %s", a.ID, clusterSocket.ClusterID))
-				continue
+			deployed, err := processDeploySocketForApp(clusterSocket, a)
+			if err != nil {
+				_, ok := appDeployLoopErrorBackoff[a.ID]
+				if !ok {
+					appDeployLoopErrorBackoff[a.ID] = &util.ErrorBackoff{MinPeriod: 1 * time.Second, MaxPeriod: 30 * time.Minute}
+				}
+				appDeployLoopErrorBackoff[a.ID].OnError(err, func() {
+					logger.Error(errors.Wrapf(err, "failed to run deploy loop for app %s in cluster %s", a.ID, clusterSocket.ClusterID))
+				})
+			} else if deployed {
+				logger.Infof("Deploy success for app %s in cluster %s", a.ID, clusterSocket.ClusterID)
 			}
 		}
+
+		time.Sleep(5 * time.Second)
 	}
 }
 
-func processDeploySocketForApp(clusterSocket *ClusterSocket, a *apptypes.App) error {
-	if a.RestoreInProgressName != "" {
-		return nil
+func processDeploySocketForApp(clusterSocket *ClusterSocket, a *apptypes.App) (bool, error) {
+	deployedVersion, err := getDeployedVersionForApp(clusterSocket, a)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get deployed version")
+	} else if deployedVersion == nil {
+		return false, nil
 	}
 
-	deployedVersion, err := downstream.GetCurrentVersion(a.ID, clusterSocket.ClusterID)
-	if err != nil {
-		return errors.Wrap(err, "failed to get current downstream version")
+	if value, ok := clusterSocket.LastDeployedSequences[a.ID]; ok && value == deployedVersion.ParentSequence {
+		// this version is already the currently deployed version
+		return false, nil
 	}
 
-	if deployedVersion == nil {
-		return nil
+	if err := deployVersionForApp(clusterSocket, a, deployedVersion); err != nil {
+		return false, errors.Wrap(err, "failed to deploy version")
 	}
+	return true, nil
+}
 
-	if value, ok := clusterSocket.LastDeployedSequences[a.ID]; ok && value == deployedVersion.ParentSequence {
-		// this version is already the currently deployed version
-		return nil
+func getDeployedVersionForApp(clusterSocket *ClusterSocket, a *apptypes.App) (*downstreamtypes.DownstreamVersion, error) {
```

Comment:
i don't understand why this function is needed. also, returning `nil` if there's a restore in progress seems unintuitive given the function name.

Reply:
i removed the extra function